<launch>
    <arg name="use_gui" default="false"/>
    <arg name="use_rviz" default="false"/>
    <arg name="robot_type" value="$(env ROBOT_TYPE)" doc="robot_type [X1,X3,X3plus,R2,X7]"/>

    <!-- ???? -->
    <!-- Simulation model -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder -->
        '$(find yahboomcar_description)/urdf/yahboomcar_$(arg robot_type).urdf'"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <!--    <node name="joint_state_publisher" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui"-->
    <!--          if="$(arg use_gui)"/>-->
    <!--    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"-->
    <!--          unless="$(arg use_gui)"/>-->

    <!-- ?????? -->
    <!-- Low-level driver node -->

    <node pkg="car_bringup" type="Mcnamu_X3plus.py" name="driver_node" required="true" output="screen" if="$(eval arg('robot_type') == 'X3plus')">
        <param name="xlinear_speed_limit" type="double" value="0.7"/>
        <param name="ylinear_speed_limit" type="double" value="0.7"/>
        <param name="angular_speed_limit" type="double" value="3.2"/>
        <param name="imu_link" type="string" value="imu_link"/>
        <remap from="/pub_vel" to="/vel_raw"/>
        <remap from="/pub_imu" to="/imu/data_raw"/>
        <remap from="/pub_mag" to="/mag/mag_raw"/>
    </node>

    <!-- ??????? -->
    <!-- Publish odometer data -->
    <node pkg="car_bringup" type="base_node" name="odometry_publisher">
        <param name="odom_frame" type="string" value="odom"/>
        <param name="base_footprint_frame" type="string" value="base_footprint"/>
        <param name="linear_scale_x" type="double" value="1.1"/>
        <param name="linear_scale_y" type="double" value="0.95"/>
        <remap from="/sub_vel" to="/vel_raw"/>
        <remap from="/pub_odom" to="/odom_raw"/>
    </node>

    <!-- ?????imu?? -->
    <!-- Filter and fuse imu data -->
    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_madgwick" output="screen">
        <param name="fixed_frame" value="base_link"/>
        <param name="use_mag" value="false"/>
        <param name="publish_tf" value="false"/>
        <param name="use_magnetic_field_msg" value="false"/>
        <param name="world_frame" value="enu"/>
        <param name="orientation_stddev" value="0.05"/>
        <param name="angular_scale" value="1.03"/>
        <remap from="/sub_imu" to="/imu/imu_raw"/>
        <remap from="/sub_mag" to="/mag/mag_raw"/>
        <remap from="/pub_imu" to="/imu/imu_data"/>
        <remap from="/pub_mag" to="/mag/mag_field"/>
    </node>

    <!-- ????????? -->
    <!-- Extended Kalman data fusion -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" output="screen">
        <remap from="odometry/filtered" to="odom"/>
        <remap from="/imu0" to="/imu/data"/>
        <remap from="/odom0" to="odom_raw"/>
        <param name="odom_frame" value="/odom"/>
        <param name="world_frame" value="/odom"/>
        <param name="base_link_frame" value="/base_footprint"/>
        <rosparam command="load" file="$(find car_bringup)/param/robot_localization.yaml"/>
    </node>

    <!-- ?????? -->
    <!-- Handle control node -->
    <include file="$(find car_ctrl)/launch/joy.launch"/>
    <!-- rviz -->
    <node pkg="rviz" type="rviz" name="odom_rviz" required="true"
          args="-d $(find car_bringup)/rviz/odom.rviz" if="$(arg use_rviz)"/>
</launch>
